### PR TITLE
[Relay] Flexible shape dispatch transformation

### PR DESCRIPTION
**Pull request by jwfromm**
This PR adds a new pass to `relay.transform` that creates a dispatcher around an input module to handle multiple input shapes. For example consider a case where I'd like to optimize my model to handle both `batch_size=1` and `batch_size=4`. I can now do so elegantly as follows:
```
shape_dict = {'data': [1, 3, 224, 224]}
model_bs1 = tvmc.load('my_model.onnx', shape_dict=shape_dict)
tvmc.tune(model, log_file='batch_1.logs')

shape_dict = {'data': [4, 3, 224, 224]}
model_bs4 = tvmc.load('my_model.onnx', shape_dict=shape_dict)
tvmc.tune(model, log_file='batch_4.logs')

# Create dispatcher for multiple batch sizes.
flex_mod = relay.transform.FlexibleShapeDispatch(buckets=[1, 4])(model_bs1.mod)

with ApplyHistoryBest(['batch_1.logs', 'batch_4.logs']):
    exe = relay.vm.compile(flex_mod, "llvm")

# Now we can run inputs with either batch 1 or batch 4 and get the tuned performance!
batch_1 = np.random.rand(1, 3, 224, 224).astype("float32")
vm.benchmark(tvm.cpu(), batch_2, func_name="main")

batch_4 = np.random.rand(4, 3, 224, 224).astype("float32")
vm.benchmark(tvm.cpu(), batch_4, func_name="main")
```

As seen above `FlexibleShapeDispatch` is a simple halfway point between fully static and fully dynamic graphs that allows us to leverage TVM tuning. If an input shape is not provided in `buckets`, it will either run fully dynamically using `relay.Any`, or if the `auto_pad` argument is set for `FlexibleShapeDispatch`, padding will be applied to match the closest bucket.

There are a few special cases that this pass handles. Multiple dynamic inputs (like those you might see in BERT) can be handled by setting `input_indices` to indicate which inputs have a dynamic axis. `affects_output` can be set to `False` for cases where the output shape is not dependent on input dynamism which could occur in dynamic resolution cases or something.

To make applying tuning logs more convenient, I also added the ability to load and merge multiple files to both autotvm and autoscheduler.

Thanks @jroesch for providing the backbone of this implementation.

### CODE REVIEW

File: python/tvm/relay/transform/flexible_shape.py [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+# pylint: disable=invalid-name, dangerous-default-value
+"""Relay functions for wrapping a module with flexible shape dispatch."""
+from tvm import relay
+
+
+def override_shape(tensor_type, dim, value):
+    """Change a value in a tensor shape."""
```
Comment (jroesch) [inline python/tvm/relay/transform/flexible_shape.py:23]:
```suggestion
    """Change a dimension in a tensor shape."""
```

File: python/tvm/relay/transform/flexible_shape.py [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+    new_dims = list(tensor_type.shape)
+    new_dims[dim] = value
+    return relay.TensorType(new_dims, tensor_type.dtype)
+
+
+def specialize_body(mod, function, dim, value, input_indices=[0], affects_output=True):
+    """Create a subgraph to handle specific input shapes"""
```
Comment (jroesch) [inline python/tvm/relay/transform/flexible_shape.py:30]:
```suggestion
    """Specialize a function to specific input shapes. This function has a few modes XXX"""
```

Would be worth explaining this a little more.

File: python/tvm/relay/transform/flexible_shape.py [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+
+
+def override_shape(tensor_type, dim, value):
+    """Change a value in a tensor shape."""
+    new_dims = list(tensor_type.shape)
+    new_dims[dim] = value
+    return relay.TensorType(new_dims, tensor_type.dtype)
+
+
+def specialize_body(mod, function, dim, value, input_indices=[0], affects_output=True):
```
Comment (jroesch) [inline python/tvm/relay/transform/flexible_shape.py:29]: You probably don't want to use the default for input_indices here, default values in Python are allocated precisely once, if you accidentally mutate this anywhere the mutations will be visible by *all invocations* of this function.

File: python/tvm/relay/transform/flexible_shape.py [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+    gvar = relay.GlobalVar("main_" + str(value))
+    # Add the new function to the main IRModule.
+    mod[gvar] = relay.Function(
+        new_params, new_body, new_ret_ty, function.type_params, function.attrs
+    )
+    return gvar, [d.type_annotation for d in dyn_data_array]
+
+
+def flexible_dispatch(
+    mod, dim=0, buckets=[1], auto_pad=False, pad_value=0, input_indices=[0], affects_output=True
```
Comment (jroesch) [inline python/tvm/relay/transform/flexible_shape.py:59]: Same comment as above, probably worth specifying the modes in the doc string if possible to enumerate.

File: python/tvm/relay/transform/flexible_shape.py [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+    Returns
+    -------
+    ret : FlexibleShapeDispatch
+        A pass that can be applied to a module to add flexible shape handling.
+    """
+
+    def __init__(
+        self,
+        dim=0,
+        buckets=[1],
```
Comment (jroesch) [inline python/tvm/relay/transform/flexible_shape.py:248]: same as above.

**Review by jroesch — APPROVED** [review]
LGTM modulo feedback

**Comment by jroesch** [comment]
cc @mbs-octoml

**Comment by jwfromm** [comment]
@jroesch I think the documentation should now be substantially improved based on your input. Let me know what you think.

**Review by mbs-octoml — APPROVED** [review]
LGTM, some optional nits:
 - consider asserting dim is Any in override_shape to avoid surprises
 - rename dim -> axis, value -> dim
 - should we force a TypeInfer afterwards to be sure the Any->dim subst is sound?

**Comment by jwfromm** [comment]
All feedback has been addressed and tests are green. @jroesch do you want to give this one more pass?

**Pull request merged**